File: Assignment2/1.2_Section_2/code.py
# Closures over a loop variable: late binding
funcs = []
for i in range(5):
    funcs.append(lambda: print(i))  # captures name i (evaluated at call time)

for f in funcs:
    f()  # prints: 4 4 4 4 4

# Fix: snapshot i via default-arg
funcs2 = [(lambda i=i: print(i)) for i in range(5)]
for f in funcs2:
    f()  # prints: 0 1 2 3 4

File: Assignment2/1.2_Section_2/code.js
// Problem with function-scoped var (old style)
var funcs = [];
for (var i = 0; i < 5; i++) {
  funcs.push(function () { console.log(i); });
}
funcs.forEach(f => f()); // 5 5 5 5 5

// Fix: block-scoped let creates a per-iteration binding
let funcs2 = [];
for (let j = 0; j < 5; j++) {
  funcs2.push(() => console.log(j));
}
funcs2.forEach(f => f()); // 0 1 2 3 4